File: with-lattice/RPS-model/src/rps.c
#include "../rps.h"

void ic(const gsl_rng *, int *); /* initial conditions	*/
void op(int , int *);		   /* data output        */                

int main(int argc, char **argv){
	int i, j, t, n, nb, gd, ac, ps, temp; /* i, j: index; 
								 	 t: time; 
								 	 n: number of files in the output
								 	 ac: active individual; 
								 	 ps: passive individual; 
								 	 temp: variable to store a temporary value; 
								 	 gd: generations; 
								 	 nb: neighbor */
	int *phi;  /* phi: vector that stores which species/empty space is at a given place on the network */
	double at; /* action to be taken */

	phi= (int *) calloc(Nx*Ny, sizeof(int));

	gsl_rng_default_seed= (argc == 2) ? atoi(argv[1]) : time(NULL);
	gsl_rng *w= gsl_rng_alloc(gsl_rng_taus);

	n= 0;
	ic(w, phi);
	op(n++, phi);

	for(t= 1; t<= NG; t++){
		gd= 0;
		while(gd < Nx*Ny){
			i= gsl_rng_uniform(w)*Nx;
			j= gsl_rng_uniform(w)*Ny;
			ac= j*Nx+i;
			if(phi[ac] != 0){
				nb= gsl_rng_uniform(w)*4;
				switch(nb){ /* neighbor selection with continuous boundary conditions */ 
					case 0:
						ps= ((j+1)%Ny)*Nx+i;
					break;
					case 1:
						ps= ((j-1+Ny)%Ny)*Nx+i;
					break;
					case 2:
						ps= j*Nx+(i+1)%Nx;
					break;
					default:
						ps= j*Nx+(i-1+Nx)%Nx;
				}
				at= gsl_rng_uniform(w);
				if(at < pm){ /* mobility */
					temp= phi[ac];
					phi[ac]= phi[ps]; 
					phi[ps]= temp;
					gd++;
				}else{
					if(at > pm && at < (pm+pp)){
						if((phi[ac]%NS) == (phi[ps]-1)){ /* predation */
							phi[ps]= 0;
							gd++;
						}
					}else{
						if(phi[ps] == 0){
							phi[ps]= phi[ac]; /* reproduction */
							gd++;
						}
					}
				}
			}
		}
		if(t % 100 == 0){ /* printing of snapshots */
			op(n++, phi);
		}
	}

	free(phi);
	gsl_rng_free(w);
	return 0;
}

File: with-lattice/RPS-model_go/rps.go
//author      		    : renata <renatatrintin@gmail.com>
//notes       		    : RPS_model with lattice
//compilation/execution : go run rps.go *seed

package main

import (
	"fmt"
	"math/rand"
	"log"
	"os"
	"strconv"
)

func main() {
	dx := 250  /* grid size      		*/
	dy := 250  /* grid size      		*/
	N  := 1000 /* number of generation	*/
	seed, _ := strconv.ParseInt(os.Args[1], 10, 64)
	rand.Seed(seed)

	phi := make([]int, dx*dy) /* phi: vector that stores which species/empty space is at a given place on the network */
//	phi := make([][]int, dx)
//	for i := 0; i < dx; i++ {
//		phi[i] = make([]int, dy)
//	}
	dst := make([]int, 4)     /* density species time */

/* INITIAL CONDITION - begin  */
	for i := 0; i < dx; i++ {
		for j := 0; j < dy; j++ {
			b := rand.Intn(3)
			phi[j*dx+i] = b+1
		}
	}
	
	f, err := os.Create("dat/data-0.dat")
	if err != nil {
     	log.Fatal(err)
	}
	
	for i := 0; i < dx; i++ {
		for j := 0; j < dy; j++ {
			fmt.Fprint(f, phi[i*dy+j], " ")
		}
		fmt.Fprintln(f, "")
	}
	f.Close()
/* FINAL CONDITION - end  */
	for i := 0; i < dx*dy; i++{
		dst[phi[i]]++
	}

	h, _ := os.OpenFile("dat/dst.dat", os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	for i := 1; i < 4; i++ { /* printing of initial species density over time */
		fmt.Fprintf(h, "%e ", float64(dst[i])/float64((dx*dy)))
	}
	fmt.Fprintf(h, "%e\n", float64(dst[0])/float64((dx*dy)))
	defer h.Close()
	
	for t := 0; t < N; t++ {
		gd := 0
		for gd < (dx*dy){
			k := rand.Intn(dx) /* line   */
			l := rand.Intn(dy) /* column */
			ac := k*dy+l
			if phi[ac] != 0 {
				ps := 5
				viz := rand.Intn(4)
				switch viz{
					case 0:
        					ps = k*dy+((l+1)%dy)
					case 1:
        					ps = k*dy+((l-1+dy)%dy)
    					case 2:
        					ps = ((k+1)%dx)*dy+l
    					case 3:
        					ps = ((k-1+dx)%dx)*dy+l
    				}
    				action := rand.Intn(100)
    				if action < 50 { /* mobility */
					temp := phi[ps];
					phi[ps]= phi[ac];
					phi[ac]= temp;
					gd++;
				}else if action < (50+25){ /* predation */
					if phi[ps] != 0 && phi[ac]%3 == phi[ps]-1{ 
						dst[phi[ps]]--;
						dst[0]++;
						phi[ps]= 0;
						gd++;
					}	
				}else if phi[ps] == 0{ /* reproduction */
					phi[ps]= phi[ac];
					dst[phi[ps]]++;
					dst[0]--;
					gd++;
				}
			}
		}
		for i := 1; i < 4; i++ { /* printing of species density over time */
			fmt.Fprintf(h, "%e ", float64(dst[i])/float64((dx*dy)))
		}
		fmt.Fprintf(h, "%e\n", float64(dst[0])/float64((dx*dy)))
		
		if t%100 == 0{ /* printing of snapshots */
			n := t/100;
			name := fmt.Sprintf("dat/data-%d.dat", n+1)
			fmt.Println(name, t)
			g, _ := os.Create(name)
			for i := 0; i < dx; i++ {
				for j := 0; j < dy; j++ {
					fmt.Fprint(g, phi[i*dy+j], " ")
				}
				fmt.Fprintln(g, "")
			}
			g.Close()
		}	
	}	
	fmt.Println("done")
}

